packages/react/src/JsonFormsContext.tsx

```diff
@@ -208,36 +407,10 @@ export const JsonFormsStateProvider = ({
     ]
   );
 
-  const onChangeRef = useRef(onChange);
   useEffect(() => {
-    onChangeRef.current = onChange;
-  }, [onChange]);
-
-  /**
-   * A common pattern for users of JSON Forms is to feed back the data which is emitted by
-   * JSON Forms to JSON Forms ('controlled style').
-   *
-   * Every time this happens, we dispatch the 'updateCore' action which will be a no-op when
-   * the data handed over is the one which was just recently emitted. This allows us to skip
-   * rerendering for all normal cases of use.
-   *
-   * However there can be extreme use cases, for example when using Chrome Auto-fill for forms,
-   * which can cause JSON Forms to emit multiple change events before the parent component is
-   * rerendered. Therefore not the very recent data, but the previous data is fed back to
-   * JSON Forms first. JSON Forms recognizes that this is not the very recent data and will
-   * validate, rerender and emit a change event again. This can then lead to data loss or even
-   * an endless rerender loop, depending on the emitted events chain.
-   *
-   * To handle these edge cases in which many change events are sent in an extremely short amount
-   * of time we debounce them over a short amount of time. 10ms was chosen as this worked well
-   * even on low-end mobile device settings in the Chrome simulator.
-   */
-  const debouncedEmit = useCallback(
-    debounce((...args: any[]) => onChangeRef.current?.(...args), 10),
-    []
-  );
-  useEffect(() => {
-    debouncedEmit({ data: core.data, errors: core.errors });
+    if (initState.config.mode !== 'controlled') {
```

Comment:
Having mode checking logic still be needed in `JsonFormsDataProvider` (and not be in the state providers) may be indicative of an opportunity to improve things further.